### PR TITLE
fix "UserWarning: Creating a tensor from a list of numpy.ndarrays is …

### LINKED ISSUE #24764: UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow
### System Info

System Info
nvidia CUDA Version: 12.1,  Driver Version: 525.105.17
transformers-cli env is
- `transformers` version: 4.29.2
- Platform: Linux-4.19.87-netease6-1-x86_64-with-glibc2.29
- Python version: 3.8.10
- Huggingface_hub version: 0.14.1
- Safetensors version: not installed
- PyTorch version (GPU?): 2.1.0a0+fe05266 (True)

### Who can help?

_No response_

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

when i use tokenizer(...,padding=True), i will get a warnning : "UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor."  and this function latency is 0.744675874710083  if query_list over 800 words .
```
st = time.time()
inputs = tokenizer(query_list, return_tensors="pt" ,padding=True)
print(time.time() - st)
```

### Expected behavior

Reduce latency and fix the  UserWarning.

### PR DESCRIPTION
**Pull request by liucw2012**
fix "UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor."

# What does this PR do?

reduce latency of codes below from 0.744675874710083s to 0.013312816619873047s.

Fixes #24764 

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker

### CODE REVIEW

File: src/transformers/tokenization_utils_base.py [hunk 1]
```diff
@@ -700,8 +700,11 @@ def convert_to_tensors(
                 raise ImportError("Unable to convert output to PyTorch tensors format, PyTorch is not installed.")
             import torch
 
-            as_tensor = torch.tensor
             is_tensor = torch.is_tensor
+
+            def as_tensor(value, dtype=None):
+                return torch.tensor(np.array(value))
+
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:707]: Do we also gave to handle case of a ragged lists as it is done with jax? Also wondering if this should always be done or not.  In this case we have a ` list of numpy.ndarrays ` but what if we have a ` numpy.ndarrays ` or a list of list, does adding a conversion have an impact? 
It might be a total nit, but I'm just wondering if we should just add a check to do it

**Review by ArthurZucker — COMMENTED** [review]
Thanks a lot for the catch and proposing a fix!

File: src/transformers/tokenization_utils_base.py (hunk 1 above)
Comment (liucw2012) [reply src/transformers/tokenization_utils_base.py:707]: I see.  but i'm not quit sure it's easy to do.

**Comment by liucw2012** [comment]
@ArthurZucker   Do I need to make any changes to this PR? What's the next step?

**Comment by liucw2012** [comment]
@ydshieh  could review this pr?

**Comment by ydshieh** [comment]
Seems to have ~30x speed up.

```python
import numpy as np
import torch

import time

def measure(batch_size, seq_len):

    a = np.ones(shape=(batch_size, seq_len, 16))
    # A list of numpy arrary
    b = [x for x in a]

    # directly to torch.tensor
    st = time.time()
    c = torch.tensor(b)
    t1 = time.time() - st

    # np -> tensor
    st = time.time()
    d = np.array(b)
    e = torch.tensor(d)
    t2 = time.time() - st

    print(f"batch_size: {batch_size} | seq_len: {seq_len} | main: {t1} sec. | PR: {t2} sec.")


batch_size = 128
seq_len = 32
for idx in range(10):
    batch_size = batch_size * 2
    measure(batch_size, seq_len)


batch_size = 128
seq_len = 256
for idx in range(8):
    batch_size = batch_size * 2
    measure(batch_size, seq_len)
```


results:

```bash
batch_size: 256 | seq_len: 32 | main: 0.010269403457641602 sec. | PR: 0.002008676528930664 sec.
batch_size: 512 | seq_len: 32 | main: 0.015998125076293945 sec. | PR: 0.0010027885437011719 sec.
batch_size: 1024 | seq_len: 32 | main: 0.03223681449890137 sec. | PR: 0.0019538402557373047 sec.
batch_size: 2048 | seq_len: 32 | main: 0.0663607120513916 sec. | PR: 0.004067182540893555 sec.
batch_size: 4096 | seq_len: 32 | main: 0.13183259963989258 sec. | PR: 0.0060040950775146484 sec.
batch_size: 8192 | seq_len: 32 | main: 0.26061558723449707 sec. | PR: 0.011055707931518555 sec.
batch_size: 16384 | seq_len: 32 | main: 0.5237565040588379 sec. | PR: 0.02300405502319336 sec.
batch_size: 32768 | seq_len: 32 | main: 1.0568530559539795 sec. | PR: 0.041966915130615234 sec.
batch_size: 65536 | seq_len: 32 | main: 2.0813064575195312 sec. | PR: 0.0868995189666748 sec.
batch_size: 131072 | seq_len: 32 | main: 4.243735074996948 sec. | PR: 0.17353129386901855 sec.
```

```bash

batch_size: 256 | seq_len: 256 | main: 0.06456398963928223 sec. | PR: 0.0034742355346679688 sec.
batch_size: 512 | seq_len: 256 | main: 0.12811279296875 sec. | PR: 0.005001068115234375 sec.
batch_size: 1024 | seq_len: 256 | main: 0.26175403594970703 sec. | PR: 0.010001659393310547 sec.
batch_size: 2048 | seq_len: 256 | main: 0.5197086334228516 sec. | PR: 0.019011259078979492 sec.
batch_size: 4096 | seq_len: 256 | main: 1.040560245513916 sec. | PR: 0.03655409812927246 sec.
batch_size: 8192 | seq_len: 256 | main: 2.089771032333374 sec. | PR: 0.07351517677307129 sec.
batch_size: 16384 | seq_len: 256 | main: 4.197775602340698 sec. | PR: 0.1453232765197754 sec.
batch_size: 32768 | seq_len: 256 | main: 8.368194103240967 sec. | PR: 0.36582493782043457 sec.
```

**Review by ydshieh — APPROVED** [review]
Verified with a code snippet provided in a comment. Indeed good! Nice

File: src/transformers/tokenization_utils_base.py (hunk 1 above)
Comment (ydshieh) [reply src/transformers/tokenization_utils_base.py:707]: @sgugger 

As the previous code is just `as_tensor = torch.tensor`, we don't need to worry about the different input types `value` here.

The only thing I don't like much is it now takes an extra line to see we are doing is `torch.tensor`. But as it indeed make things faster (~30x), I am good.

Is this PR good for you?

File: src/transformers/tokenization_utils_base.py [hunk 2]
```diff
@@ -700,8 +700,11 @@ def convert_to_tensors(
                 raise ImportError("Unable to convert output to PyTorch tensors format, PyTorch is not installed.")
             import torch
 
-            as_tensor = torch.tensor
             is_tensor = torch.is_tensor
+
+            def as_tensor(value, dtype=None):
+                return torch.tensor(np.array(value))
```
Comment (sgugger) [inline src/transformers/tokenization_utils_base.py:706]: Could we just check here if the `value` is a list of `numpy.ndarray` since this is the case we have to fix, and simply use `torch.tensor` in the other cases?

Comment (ydshieh) [reply src/transformers/tokenization_utils_base.py:706]: Nice suggestion!

Comment (liucw2012) [reply src/transformers/tokenization_utils_base.py:706]: it's also works to list. 
```
import numpy as np
import torch
import random
import time

def measure():   
    randomlist = []
    for i in range(0, 1024000):
        a = random.randint(1,100)
        randomlist.append(a)
    # list -> tensor
    st = time.time()
    e = torch.tensor(randomlist)
    t3 = time.time() - st

    # list -> np -> tensor
    st = time.time()
    f = torch.tensor(np.array(randomlist))
    t4 = time.time() - st

    print(f"list->tensor: {t3} sec. | list->np->tensor: {t4} sec.")



for idx in range(10):
    measure()
```
results:
```
list->tensor: 0.11251711845397949 sec. | list->np->tensor: 0.08889651298522949 sec.
list->tensor: 0.11182880401611328 sec. | list->np->tensor: 0.056847572326660156 sec.
list->tensor: 0.11099457740783691 sec. | list->np->tensor: 0.05408644676208496 sec.
list->tensor: 0.11319661140441895 sec. | list->np->tensor: 0.052817344665527344 sec.
list->tensor: 0.11264157295227051 sec. | list->np->tensor: 0.05298781394958496 sec.
list->tensor: 0.10972905158996582 sec. | list->np->tensor: 0.05270671844482422 sec.
list->tensor: 0.10953164100646973 sec. | list->np->tensor: 0.05267834663391113 sec.
list->tensor: 0.10898041725158691 sec. | list->np->tensor: 0.053290367126464844 sec.
list->tensor: 0.10944771766662598 sec. | list->np->tensor: 0.05272650718688965 sec.
list->tensor: 0.11078786849975586 sec. | list->np->tensor: 0.0529322624206543 sec.
```

Comment (liucw2012) [reply src/transformers/tokenization_utils_base.py:706]: actually. In my case,  i use tokenizer with padding is true. it will padding all the queries to the longest one.  i got almost 5ms latency reduced if one query is 800 Chinese characters.

Comment (sgugger) [reply src/transformers/tokenization_utils_base.py:706]: Yes, but you are not the only person using the Transformers library. For smaller sizes, it's faster to just call `torch.tensor` as is right now, so let's not change that part.

Comment (liucw2012) [reply src/transformers/tokenization_utils_base.py:706]: i can fix it if you insist.  i just share my experience. not every one  have the chance or ability to make their model faster in several ms.

**Review by sgugger — APPROVED** [review]
Thanks for bearing with me!
